direct/src/gui/DirectFrame.py

```diff
@@ -50,166 +52,113 @@ def __init__(self, parent = None, **kw):
-        if geom is not None:
-            self['geom'] = geom
-
-        # Determine argument type
-        geom = self['geom']
-
-        if geom == None:
-            # Passed in None
-            geomList = (None,) * self['numStates']
-        elif isinstance(geom, NodePath) or \
-             isinstance(geom, stringType):
-            # Passed in a single node path, make a tuple out of it
-            geomList = (geom,) * self['numStates']
+    def long_type_check(self, name, object_):
+        if isinstance(object_, stringType):
+            return True
+        if name == "image":
+            if ((isinstance(object_, NodePath) or
+                isinstance(object_, Texture)) 
+                or
+                ((len(object_) == 2) and# Its a model/node pair of strings
+                isinstance(object_[0], stringType) and
+                isinstance(object_[1], stringType))):
+                return True
+        elif name == "geom":
+            if (isinstance(object_, NodePath)):
+                return True
+        return False
+
+    def new_set(self, name, object_=None):
+        """my new code golf par is 1"""
+        assert name in ["geom", "image", "text"]
+        if object_ is not None:
+            self[name] = object_
+
+        object_ = self[name]
+
+        if object_ == None:
+            object_list = (None,) * self['numStates']
         else:
-            # Otherwise, hope that the user has passed in a tuple/list
-            geomList = geom
-
-        # Create/destroy components
-        for i in range(self['numStates']):
-            component = 'geom' + repr(i)
-            # If fewer items specified than numStates,
-            # just repeat last item
-            try:
-                geom = geomList[i]
-            except IndexError:
-                geom = geomList[-1]
-
-            if self.hascomponent(component):
-                if geom == None:
-                    # Destroy component
-                    self.destroycomponent(component)
+            if self.long_type_check(name, object_):
+                object_list = (object_,) * self['numStates']
+            else:
+                object_list = object_
+
+        name_based_classes = {
+            "geom": OnscreenGeom,
+            "image": OnscreenImage,
+            "text": OnscreenText, }
+
+        name_based_constants = {
+            "geom": {
+                "scale": 1,
+                "sort": DGG.GEOM_SORT_INDEX, # this is "20"
+            },
+            "image": {
+                "scale": 1,
+                "sort": DGG.IMAGE_SORT_INDEX, # this is "10"
+            },
+            "text": {
+                "scale": 1,
+                "mayChange": self['textMayChange'],
+                "sort": DGG.TEXT_SORT_INDEX, # this is "30"
+            },
+        }
+        # constants should be local to or default arguments of constructors
+
+        m = self["numStates"]
+        for c in range(self['numStates']):
```

Comment:
You're querying this twice in a row, maybe you meant `for c in range(m):`?  That said, the variable should also probably be named more descriptively than a single letter.

Reply:
Hm it's the state counter. Sure we can name it state_i, but I'm not sure how necessary that is. It's only used 3 times within like 20 lines  and the loop starts with "for i in self["numstates"]". I think it's pretty obvious that that is the state number.

I will change it if you insist though.

Reply:
I'm talking about the assignment to `m`.  It makes sense to assign it to a temporary variable, but only if you're going to then use that variable. :)